client/asset/eth/eth.go

```diff
@@ -0,0 +1,306 @@
+package main
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/cmd/utils"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/downloader"
+	"github.com/ethereum/go-ethereum/eth/ethconfig"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/les"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/nat"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type ethereumI interface {
+	APIs() []rpc.API
+	Stop() error
+}
+
+type ETH struct {
+	ctx       context.Context
+	client    *ethclient.Client
+	rpcClient *rpc.Client
+	shutdown  context.CancelFunc
+	nodeCfg   *node.Config
+	node      *node.Node
+	ethereum  ethereumI
+}
+
+func NewETH(appDir string, goerli bool) (*ETH, error) {
+	stackConf := &node.Config{DataDir: appDir}
+
+	stackConf.P2P.MaxPeers = 10
+	var key *ecdsa.PrivateKey
+	var err error
+	if key, err = crypto.GenerateKey(); err != nil {
+		return nil, err
+	}
+	stackConf.P2P.PrivateKey = key
+	stackConf.P2P.ListenAddr = ":30303"
+	stackConf.P2P.NAT = nat.Any()
+
+	var urls []string
+	switch {
+	case goerli:
+		urls = params.GoerliBootnodes
+	default:
+		urls = params.MainnetBootnodes
+	}
+
+	for _, url := range urls {
+		node, err := enode.Parse(enode.ValidSchemes, url)
+		if err != nil {
+			return nil, fmt.Errorf("Bootstrap URL %q invalid: %v", url, err)
+		}
+		stackConf.P2P.BootstrapNodes = append(stackConf.P2P.BootstrapNodes, node)
+	}
+
+	for _, url := range params.V5Bootnodes {
+		node, err := enode.Parse(enode.ValidSchemes, url)
+		if err != nil {
+			return nil, fmt.Errorf("Bootstrap v5 URL %q invalid: %v", url, err)
+		}
+		stackConf.P2P.BootstrapNodesV5 = append(stackConf.P2P.BootstrapNodesV5, node)
+	}
+
+	stack, err := node.New(stackConf)
+	if err != nil {
+		return nil, err
+	}
+
+	ethCfg := ethconfig.Defaults
+	if goerli {
+		ethCfg.Genesis = core.DefaultGoerliGenesisBlock()
+		ethCfg.NetworkId = params.GoerliChainConfig.ChainID.Uint64()
+	}
+
+	// ethCfg.Genesis = core.DefaultGenesisBlock() // eth will do this anyway when Genesis == nil is encountered, and the messaging it better that way.
+	// ethCfg.SyncMode = downloader.SnapSync // Supposed to be enabled with Berlin on April 14th 2021
+	// ethCfg.SyncMode = downloader.FastSync  // The default
+	ethCfg.SyncMode = downloader.LightSync
```

Comment:
Updated here after @JoeGruffins reminded me that fast != light. Much faster now. I'm leaning towards think only enabling light mode to start, and then add snap mode option after Berlin.

Reply:
I got the impression that snap sync was going to be the *default* after Berlin, but it could actually work now if other nodes are making snapshots.  But that's aside because light is the way to go for client.